LanguageFeatures/Patterns/type_inference_A12_t02.dart

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Calculate the static type of the pattern. Using that value type,
+/// recurse through the pattern again downwards to the leaf subpatterns filling
+/// in any missing types in the pattern. This process may also insert implicit
+/// coercions and casts from dynamic when values flow into a pattern during
+/// matching.
+///
+/// @description Check that the calculation og the static type of a null-assert
+/// pattern performs casts from dynamic
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+import "../../Utils/static_type_helper.dart";
+import "../../Utils/expect.dart";
+
+main() {
+  dynamic pi = 3.14;
+  final (double? v1!) = pi;
+  v1.expectStaticType<Exactly<double?>>();
+
+  Expect.throws(() {
+    var (int? v2!) = pi as dynamic;
```

Comment:
`pi` is declared to have type `dynamic`. It is true that we're assigning it to a `double?`, and also that the `!` pattern did not throw, but `double?` and `double` are not types of interest for `pi`, and we can't compute a `NonNull(dynamic)` which is different from `dynamic`. So I'd expect `pi` to have the type `dynamic` also at this point.

Reply:
Awaits a decision in dart-lang/language#2846.

Reply:
Changed to 
```dart
  Expect.throws(() {
    dynamic e = 2.71;
    var (int? v2!) = e;
  });
```